### ISSUE TITLE
semi reports missing semicolon on export default class

### ISSUE THREAD

**Issue by btmills** [eslint/eslint#2194]
```js
// test.js
export default class {};
```

```js
// config.json
{
	"env": {
	  "es6": true
	},
	"ecmaFeatures": {
	  "modules": true
	},
    "rules": {
        "semi": 2
    }
}
```

```sh
$ eslint --reset --no-eslintrc --config config.json test.js

test.js
  2:23  error  Missing semicolon  semi

✖ 1 problem (1 error, 0 warnings)
```

Looks like similar cases were addressed in #2178 and this one was missed.

**Comment by nzakas** [comment]
Imports/exports have so many variations, it's crazy. :)

**Comment by pola88** [comment]
I have a problem with ```export``` as well.
My problem is the following:

```
export class MyObject extends Object {
  subpageOf(other) {
    return false;
  }
}
```

and I get ```error  Missing semicolon      semi```

if I add the semicolon, returns:

```
error  Missing semicolon      semi
error  Unnecessary semicolon  no-extra-semi
```

Do you know how fix it?
Thanks,

**Comment by btmills** [comment]
Working on this. It's quite the bear. In the meantime, this is my interpretation of the spec, broken down by the [productions of `ExportDeclaration`](https://people.mozilla.org/~jorendorff/es6-draft.html#sec-exports):

### export * FromClause ;

Semicolon belongs to `ExportDeclaration`.

```js
export * from 'foo';
```

### export ExportClause FromClause ;

Semicolon belongs to `ExportDeclaration`.

```js
export { foo } from 'foo';
```

### export ExportClause ;

Semicolon belongs to `ExportDeclaration`.

```js
export { foo };
```

### export VariableStatement

Semicolon belongs to `VariableStatement`.

```js
export var foo;
```

### export Declaration

#### Declaration : HoistableDeclaration

No semicolon.

```js
export function foo () { }
export function* foo () { }
```

#### Declaration : ClassDeclaration

No semicolon.

```js
export class Foo () { }
```

#### Declaration : LexicalDeclaration

Semicolon belongs to LexicalDeclaration.

```js
export let foo;
export const foo = 42;
```

### export default HoistableDeclaration

No semicolon.

```js
export default function () { }
export default function* foo() { }
```

### export default ClassDeclaration

No semicolon.

```js
export default class { }
```

### export default AssignmentExpression ;

Semicolon belongs to ExportDeclaration.

```js
export default foo || bar;
export default (foo) => foo.bar();
export default foo = 42;
export default foo += 42;
```

**tl;dr:**

- `ExportAllDeclaration` may have a semicolon.
- `ExportNamedDeclaration` whose `declaration` is `null` may have a semicolon.
- `ExportNamedDeclaration` with a child `declaration` does not have a semicolon, though its child may.
- `ExportDefaultDeclaration` does not have a semicolon, though its child may.

*(end tl;dr)*

I find this interesting, [from ESTree](https://github.com/estree/estree/blob/master/es6.md#exportnameddeclaration):

```js
interface ExportNamedDeclaration <: Node {
    type: "ExportNamedDeclaration";
    declaration: Declaration | null;
    specifiers: [ ExportSpecifier ];
    source: Literal | null;
}
```

> Note: Having `declaration` populated with non-empty `specifiers` or non-null `source` results in an invalid state.

If `declaration` and (`specifiers`, `source`) are mutually exclusive, why are they the same node?

**Comment by nzakas** [comment]
Wow, nice work!

**Comment by voithos** [comment]
This appears to have had a regression, unless I'm doing something wrong. I'm running into the message using latest `gulp-eslint`, but am also able to reproduce it with `eslint` standalone.

```
// test.js
export default class {};

// config.json
{
    "env": {
      "es6": true
    },
    "ecmaFeatures": {
      "modules": true
    },
    "rules": {
        "no-extra-semi": 2
    }
}

$ eslint --version
v1.10.1
$ eslint --no-eslintrc --config config.json test.js
test.js
  1:24  error  Unnecessary semicolon  no-extra-semi

✖ 1 problem (1 error, 0 warnings)
```

**Comment by lo1tuma** [comment]
@voithos this is the expected behavior, the semicolon in your example isn't necessary.

**Comment by voithos** [comment]
@lo1tuma Ah, you're right! That's... unexpected. Thanks!
